File: Leetcode_Code/99. Recover Binary Search Tree/99.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private TreeNode lastNode, first, second;

    public void recoverTree(TreeNode root) {
        lastNode = first = second = null;
        recoverTree_dfs(root);
        int temp = first.val;
        first.val = second.val;
        second.val = temp;
    }

    private void recoverTree_dfs(TreeNode root) {
        if (root == null) return;
        recoverTree_dfs(root.left);
        if (lastNode != null && lastNode.val >= root.val) {
            if (first == null) first = lastNode;
            if (first != null) second = root;
        }
        lastNode = root;
        recoverTree_dfs(root.right);
    }
}

File: Leetcode_Code/99. Recover Binary Search Tree/99.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void recoverTree(TreeNode *root) {
        recoverTree_dfs(root);
        if (wrongNode, secondWNode) swap(wrongNode->val, secondWNode->val);
    }

private:
    TreeNode *lastNode = NULL;
    TreeNode *wrongNode = NULL;
    TreeNode *secondWNode = NULL;

    void swap(int &a, int &b) {
        int temp = a;
        a = b;
        b = temp;
    };

    void recoverTree_dfs(TreeNode *root) {
        if (!root) return;
        recoverTree_dfs(root->left);
        if (lastNode != NULL && root->val <= lastNode->val) {
            if (wrongNode == NULL) {
                wrongNode = lastNode;
            }
            if (wrongNode != NULL) {
                secondWNode = root;
            }
        }
        lastNode = root;
        recoverTree_dfs(root->right);
    }
};
